multilabel example

franchuterivera opened the pull request:
Address #754 

I am using https://www.openml.org/d/40594 because it is a good tradeoff between features and performance.

other multilabel datasets like https://www.openml.org/d/40597 (yeast) cannot achieve a good performance with the strong resource limitations of the examples.

mfeurer commented on examples/example_multilabel_classification.py:
```diff
@@ -0,0 +1,75 @@
+# Data Loading
+# ============
+
+# Using reuters multilabel dataset -- https://www.openml.org/d/40597
```
The docstring still points to yeast.

mfeurer commented on examples/example_multilabel_classification.py:
```diff
@@ -0,0 +1,75 @@
+
+# fetch openml downloads a numpy array with TRUE/FALSE strings. Re-map it to
+# integer dtype with ones and zeros
+y[y == 'TRUE'] = 1
```
Shouldn't Auto-sklearn encode the data itself?

franchuterivera replied:
We can make this happen. Today is not possible because input validation happens for automl.

Nevertheless, the estimators perform a check-targets early in the game that throws:
ValueError: classification with data of type multiclass-multioutput is not supported

I will do a push to showcase the required changes to support this.

mfeurer commented on autosklearn/data/validation.py:
```diff
@@ -533,3 +538,34 @@ def join_and_check(
+        elif not scipy.sparse.issparse(y):
+            # It should be a list or numpy array
+            # if this point was reached
+            y_array = np.array(y)
```
Could you please check if `y` is already a numpy array and do not make a copy if that's the case?

mfeurer commented on autosklearn/data/validation.py:
```diff
@@ -533,3 +538,34 @@ def join_and_check(
                              "Kindly cast your targets to a supported type."
                              )
+
+    def _is_data_valid_for_classification(self, y: np.ndarray) -> None:
```
Could you please add unit tests that trigger this function?

mfeurer commented on autosklearn/data/validation.py:
```diff
@@ -533,3 +538,34 @@ def join_and_check(
                              "Kindly cast your targets to a supported type."
                              )
+
+    def _is_data_valid_for_classification(self, y: np.ndarray) -> None:
```
How can we actually have this function triggered? If `type_of_target` detects a continuous dataset in the classification part in automl.py, it would stop Auto-sklearn. Therefore, this appears to do double work?

franchuterivera replied:
Yes, I agree type_of_target is a super set of checks, except in the case of multi label classification, where the output is interpreted as multiclass-multioutput


I can think of 2 ways to go:
1. we don't support anything that type_of_target doesn't support and request the user to comply with it. In other words, we do not encode nor handle multilabel targets of string type. This will be the cleanest solution and we align to what scikit learn support. In this example, we will have to manually encode the target.

2. We add a code to handle this specific multilabel scenario. If the target is multidimensional, we check if dtype is encodable we proceed. So this _is_data_valid_for_classification was an attempt to go with this approach, and I realize the name is confusing. It should be something like handle multilabel and have more in-depth checks.

Please let me know your preferences

franchuterivera commented on examples/example_multilabel_classification.py:
```diff
@@ -71,5 +60,10 @@
 # Get the Score of the final ensemble
 # ===================================
 
+# y_test contains TRUE/FALSE which is not supported by sklearn directly
```
I think the situation is different:
```
X, y = sklearn.datasets.fetch_openml(data_id=40594, return_X_y=True, as_frame=False)
sklearn.metrics.accuracy_score(y_test,y_test)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/chico/work/auto-sklearn_fork_doc_pandas/venv/lib/python3.6/site-packages/sklearn/metrics/_classification.py", line 185, in accuracy_score
    y_type, y_true, y_pred = _check_targets(y_true, y_pred)
  File "/home/chico/work/auto-sklearn_fork_doc_pandas/venv/lib/python3.6/site-packages/sklearn/metrics/_classification.py", line 97, in _check_targets
    raise ValueError("{0} is not supported".format(y_type))
ValueError: multiclass-multioutput is not supported
```

Which you can also see in your example:

```
true = np.array([['True', 'False'], ['False', 'False']])
sklearn.metrics.accuracy_score(true,true)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/chico/work/auto-sklearn_fork_doc_pandas/venv/lib/python3.6/site-packages/sklearn/metrics/_classification.py", line 185, in accuracy_score
    y_type, y_true, y_pred = _check_targets(y_true, y_pred)
  File "/home/chico/work/auto-sklearn_fork_doc_pandas/venv/lib/python3.6/site-packages/sklearn/metrics/_classification.py", line 97, in _check_targets
    raise ValueError("{0} is not supported".format(y_type))
ValueError: multiclass-multioutput is not supported
```

So the changes are to support encoding for 2D targets

franchuterivera commented on autosklearn/automl.py:
```diff
@@ -1061,6 +1061,18 @@ def fit(
+                     ensemble_size=None):
+
+        # Make sure that input is valid
+        # Fit the ensemble at this stage, only in the classification
```
updated the comment to be:
```
        # In the case of classification, it is possible for the user
         # to build a estimator from scratch (a estimator object that
         # has not gone through fit(), like in psmac ). If this is the case,
         # no input validator was previously fit, so we have to make sure
         # we call validate targets with is_classification=True. The super()
         # fit_ensemble can be used for regression/classification, so we
         # should not assert is_classification there.

```

franchuterivera commented on autosklearn/estimators.py:
```diff
@@ -664,7 +665,9 @@ def fit(self, X, y,
-        target_type = type_of_target(y)
+        target_type = type_of_target(
+            InputValidator().validate_target(y, is_classification=True)
+        )
```
Sorry, can you please elaborate on this comment?

mfeurer replied:
I meant that replacing:
```python
if target_type in ['multiclass-multioutput', 'continuous', 'continuous-multioutput', 'unknown',]:
    raise ValueError("classification with data of type %s is not supported" % target_type)
```
by
```python
if target_type not in ['binary', 'multiclass']:
    raise ValueError("classification with data of type %s is not supported" % target_type)
```
So we check what is allowed instead of checking what is forbidden.

mfeurer commented:
Thanks a lot for pointing out the issues with multilabel indicators. I totally don't understand this restriction I think we should then only allow integers in multilabel data to comply with scikit-learn as much as possible. What do you think about this restriction?

Then the example would demonstrate downloading a dataframe, explicitly encoding that so that users see that this has to be done manually.

We could also consider opening an issue. I currently don't see a reason why something that is binary in every output wouldn't be a multi-label indicator task. I also checked the sklearn issue, and this one here is the closest: https://github.com/scikit-learn/scikit-learn/issues/14671

franchuterivera commented:
I reverted to a previous commit and added the suggested positive query for target types.

About the multilabel indicator, looks like the definition from here https://scikit-learn.org/stable/modules/multiclass.html actually states:

 ***Formally, a binary output is assigned to each class, for every sample. Positive classes are indicated with 1 and negative classes with 0 or -1.***

So I guess it is expected? But definitely a nice feature for them to add (support for text arguments, if they are binary)...

mfeurer commented on autosklearn/estimators.py:
```diff
@@ -795,14 +797,18 @@ def fit(self, X, y,
-                           ]:
-            raise ValueError("regression with data of type %s is not"
-                             " supported" % target_type)
+        supported_types = ['continuous', 'binary', 'multiclass', 'continuous-multioutput']
```
Hey, what's the binary and multiclass doing here?

franchuterivera replied:
I originally thought it was a mistake!, but:
```
type_of_target([1.0, 0.0, 3.0])
'multiclass'
```

And then for binary, the testing expect this:
And also the testing looks for this, for example:
https://github.com/automl/auto-sklearn/blob/859fb993607bd5cfb881aca3b346cf4d1addf470/test/test_automl/test_estimators.py#L233

I am honoring the testing requirement for binary, but I don't agree with it. Should I remove it?
I do believe we need multiclass, because how above scenario treats the floats.

franchuterivera replied:
As per discussed offline, we will leave binary case (due to 1.0 and 0.0) and multiclass.

I updated the comment  to reflect this

mfeurer commented on autosklearn/estimators.py:
```diff
@@ -795,14 +797,18 @@ def fit(self, X, y,
         # Before running anything else, first check that the
         # type of data is compatible with auto-sklearn. Legal target
-        # types are: continuous, binary, multiclass.
+        # types are: continuous, continuous-multioutput, 'multiclass'
```
I think you need to update this.

franchuterivera replied:
Same here, improved the message to show we need this.

mfeurer commented on autosklearn/estimators.py:
```diff
@@ -665,13 +665,15 @@ def fit(self, X, y,
-                           ]:
-            raise ValueError("classification with data of type %s is"
-                             " not supported" % target_type)
+        supported_types = ['binary', 'multiclass', 'multilabel-indicator']
```
In the example you gave a link to scikit-learn's explanation of task types. Could you please add it to the error message here, too?

franchuterivera replied:
Understood, I will add this in the next push.

mfeurer commented on autosklearn/estimators.py:
```diff
@@ -795,14 +797,18 @@ def fit(self, X, y,
-                             " supported" % target_type)
+        supported_types = ['continuous', 'binary', 'multiclass', 'continuous-multioutput']
+        if target_type not in supported_types:
+            raise ValueError("Regression with data of type {} is "
```
Same as above, please update the error message to include a link to scikit-learn.

franchuterivera replied:
Understood, I will add it in the next push.
